r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -612,6 +611,15 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         mut cs: CS,
         other: &Self,
     ) -> Result<NonNativeFieldMulResultGadget<SimulationF, ConstraintF>, SynthesisError> {
+        let params = get_params(SimulationF::size_in_bits(), ConstraintF::size_in_bits(), SimulationF::Params::DIFFERENCE_WITH_HIGHER_POWER_OF_TWO);
+        // first check if SimulationF is pseudo-mersenne, as in this case we call the optimized
+        // multiplication function for pseudo-mersenne fields
+        if super::is_pseudo_mersenne::<SimulationF>() {
+            let res = self.mul_without_prereduce_for_pseudomersenne(cs.ns(|| "mul for pseudo-mersenne"), other, false)?;
+            // convert res to NonNativeFieldMulResultGadget to be compatible with type returned by the function
+            return Ok(NonNativeFieldMulResultGadget::from_non_native_field_gadget(&mut cs, &res, &params))
```

Comment:
One of the advantages of this optimization based on Mersenne is that it allows us to return a NonNativeFieldGadget as a result of mul instead of a NonNativeFieldMulResultGadget (on which we can only perform additions); here, however, to keep the interfaces consistent, we are still returning a NonNativeFieldMulResultGadget, obliging us to perform a reduce if we want to do another mul while it shouldn't be needed.
In future we will refactor everything so NonNativeFieldMulResultGadget won't exist anymore but, for this PR, would be nice to test if we are wasting constraints and how much.
I suggest to perform a test that does

```
let mul_result_gadget= a.mul(b)
mul_result_gadget.reduce()
mul_result_gadget.mul(c)
```
Using the function we expose to the user and another one:

```
let non_native_field_gadget = a.mul_mersenne(b)
non_native_field_gadget.mul_mersenne(c)
```

Using instead our internal function for Mersenne mul.
Let's count number of constraints in both cases and compare

Reply:
Exactly, this is what I proposed offline: for now, let us put the discrimination logic (generic/pseudo Mersenne) in the `reduce()` of mul result gadget instead (if possible). This would allow us to make use of the optimized circuit writing of the non-native group gadget, which explicitly operates with mul result gadgets before calling the reduction.